### DISCUSSION TITLE
Draft System on GraphQL

### DISCUSSION THREAD

**Discussion by Wiz-Amit** [strapi/strapi#7679] [Legacy Topics]
Hi,
I have used the Strapi article to build a blog using Angular and Strapi. Now I am trying to filter blog articles based on two properties.

1. If the publication date has passed.
2. If the status is "Public".

I have customized the query and added more filters as per [this documentation](https://strapi.io/documentation/v3.x/guides/draft.html) and it is working fine with REST API. But, how can I get this same kind filtering in the case of GraphQL?

Thanks

**Comment by derrickmehaffy** [comment]
Can you post the GraphQL query you have tried and I can help you tweak it?

**Reply by Wiz-Amit** [reply]
I don't want to modify the query. I want to restrict access to the article with `article.status != 'Public'`.

## Query (same as Strapi blog):
```js
import gql from 'graphql-tag';

const CATEGORY_ARTICLES_QUERY = gql`
  query Category($id: ID!) {
    category(id: $id) {
      id
      name
      articles {
        id
        title
        content
        image {
          url
        }
        categories {
          id
          name
        }
      }
    }
  }
`;

export default CATEGORY_ARTICLES_QUERY;
```

## Custom Article Controller (Working fine)
```js
module.exports = {
  async find(ctx) {
    let entities;

    ctx.query = {
      ...ctx.query,
    };

    if (ctx.query._q) {
      entities = await strapi.services.article.search(ctx.query);
    } else {
      entities = await strapi.services.article.find(ctx.query);
    }

    entities = entities.map(filterArticle);

    return entities.map((entity) =>
      sanitizeEntity(entity, { model: strapi.models.article })
    );
  },
  async findOne(ctx) {
    const { id } = ctx.params;

    let article = await strapi.services.article.findOne({ id });

    article = filterArticle(article);

    return sanitizeEntity(article, { model: strapi.models.article });
  },
};

function filterArticle(article) {
  return returnIf(article, [isPublic, isPublished]) ? article : null;

  function returnIf(entity, filters) {
    return filters.every((filter) => filter(entity));
  }

  function isPublic(article) {
    return article.status == "Public";
  }

  function isPublished(article) {
    return new Date().getTime() >= new Date(article.published_at).getTime();
  }
}
```

**Reply by lauriejim** [reply]
Hello! I just checked with the tech team, and it's not possible to achieve that.

And a comment about your implementation of this thing.
I suggest you play with filters - https://strapi.io/documentation/v3.x/content-api/parameters.html#filters

```js
ctx.query = {
  ...ctx.query,
  status_ne: 'Public'
};
```

**Reply by Wiz-Amit** [reply]
Thank you for your suggestion.
I have previously filtered articles by status using query (as you have shown) and it works (with REST API). But I am not able to achieve the same result in case of GraphQL.

**Reply by derrickmehaffy** [reply]
@Wiz-Amit you should be able to filter with GraphQL using the where structure like

```
query {
  articles (where: { status_ne: "Public" } ) {
    title
    body
  }
}
```

**Reply by Wiz-Amit** [reply]
Sorry for the miscommunication. I meant to filter articles (or anything) strictly from the API. Or to be precise I want to restrict access to some data.

**Reply by derrickmehaffy** [reply]
I understand your first part about filtering from the API, and as Jim has mentioned that is possible already.

Can you please clarify your second point?
